### PR TITLE
Auto-focus on new project

### LINKED ISSUE #194: Create new card? Auto-focus textarea
## Issue - Enhancement

This improves the speed/ease of capturing actions and projects.

- **Estimated effort:** 1 point ([see CONTRIBUTING.md](https://github.com/ParabolInc/action/blob/master/CONTRIBUTING.md#points-and-sizes))

### PR DESCRIPTION
**Pull request by jordanh**
Closes #194

### CODE REVIEW

File: src/client/client.js [hunk 1]
```diff
@@ -49,6 +45,9 @@ const createCashay = (store, cashaySchema) => {
     // ENABLE THIS FOR EXPLORING FRONT END PERFORMANCE
     // const {whyDidYouUpdate} = require('why-did-you-update');
     // whyDidYouUpdate(React);
+    // ENABLE THIS AS ANOTHER MEANS OF TESTING FRONT END PERFORMANCE
+    // const Perf = require('react-addons-perf');
+    // window.Perf = Perf;
```
Comment (jordanh) [inline src/client/client.js]: I am seeing some performance degradation on occasion. I still don't know what's causing it. I appears unrelated to this change.

File: src/universal/components/OutcomeCard/OutcomeCard.js [hunk 2]
```diff
@@ -121,6 +122,9 @@ export default class OutcomeCard extends Component {
           }
         };
         cashay.mutate('updateTask', options);
+        if (autoFocus) {
+          dispatch(editingClearFocus());
```
Comment (jordanh) [inline src/universal/components/OutcomeCard/OutcomeCard.js]: Our app has a single duck for controlling what requests focus from the UI. If focus was set on a card, it gets cleared on submit.

File: src/universal/components/OutcomeCard/OutcomeCardTextarea.js [hunk 3]
```diff
@@ -92,7 +94,12 @@ export default class OutcomeCardTextAreaField extends Component {
       input.onBlur();
     };
     let textAreaRef;
-    const setRef = (c) => { textAreaRef = c; };
+    const setRefAndFocus = (c) => {
+      textAreaRef = c;
+      if (c && autoFocus) {
+        c.focus();
```
Comment (jordanh) [inline src/universal/components/OutcomeCard/OutcomeCardTextarea.js]: Here is where the focus is actually requested for the component.

File: src/universal/modules/teamDashboard/components/ProjectColumn/ProjectColumn.js [hunk 4]
```diff
@@ -38,6 +41,7 @@ const ProjectColumn = (props) => {
       userSort: 0
     };
     cashay.mutate('createTask', {variables: {newTask}});
+    dispatch(editingSetFocus(normalizedProjectId));
```
Comment (jordanh) [inline src/universal/modules/teamDashboard/components/ProjectColumn/ProjectColumn.js]: And here is where focus is requested for the new object.

**Comment by jordanh** [comment]
Lint is unrelated to these changes.

**Comment by jordanh** [comment]
I started implementing this naïvely, where autofocus was the domain of individual components. ThenI started to think through cases where the user clicks on "new action" then "new project" and our components fight for autofocus. We needed something to mediate the global state, so, I put it in a duck.

It was related enough to editing that in it went...

After a little refactoring, we now have a system for requesting and setting focus.

File: src/client/client.js (hunk 1 above)
Comment (mattkrick) [inline src/client/client.js]: this app is WILDLY unoptimized, it's true. And I know exactly how to speed it up. I have the BEST ways to speed it up, ask anyone. I just haven't worried about that unless I see a screen flicker.

Comment (jordanh) [inline src/client/client.js]: Everybody knows that only I can optimize this application. When I start to optimize this application, it's going to move so fast it's going to make the user's head spin. Don't listen to what other people might tell you. Those people are stupid. Hey, is that a baby? Would somebody through that baby outta here?

**Comment by jordanh** [comment]
Shall we merge this sucker back to `project-autofocus` @mattkrick ? I'd hate for dev to get too far ahead. Justed wanted to make sure you were cool with the approach before I plunked it in...

**Comment by mattkrick** [comment]
Gimme 30 more mins before you do, I think there might be a better way...

**Comment by mattkrick** [comment]
K, so here's how I'd simplify:
In OutcomeCard, calling the field:
```
doFocus={!content}
```

In OutcomeCardTextAreaField:
```
autoFocus={doFocus}
```

I have it in the meeting-update branch, can make it a separate PR if you like it.

**Comment by mattkrick** [comment]
Worth noting that this doesn't just work on new cards, it works on abandoned new cards, too. So if you create a card, then your browser crashes, then you restart, it'll be flashin at ya tellin ya to fill it out.

**Comment by jordanh** [comment]
Closing in favor of @mattkrick's implementation on the branch.
